### PR TITLE
gh-94499 Add test for private name mangling in class pattern matching

### PR DESCRIPTION
**Pull request by da-woods**
The current status quo is that private attribute names are not
mangled when a class is matched. I've added a test to
document/legitimize this behaviour.

I don't think this needs a news entry (but can create one if required).

I asked about [this in Python-dev](https://mail.python.org/archives/list/python-dev@python.org/thread/LILLO3MBTVY6ZQT3VNUVXATEPS3ASGQF/) and it wasn't clear if the current behaviour was by design or not. My view is that it seems reasonable so the right thing to do is to "lock it in" by testing for it, but I'm also happy for

* Issue: gh-94499

### CODE REVIEW

File: Lib/test/test_patma.py [hunk 1]
```diff
@@ -2654,6 +2654,21 @@ def get(key, default=None):
...
+            def f(self, x):
+                match x:
+                    # looks up __attr, not _C__attr or _Outer__attr
+                    case C(__attr=y):
+                        return y
+        c = C()
+        setattr(c, "__attr", 1)
+        setattr(c, "_Outer__attr", 2)
+        setattr(c, "_C__attr", 3)
+        self.assertEqual(Outer().f(c), 1)
```
Comment (brandtbucher) [inline Lib/test/test_patma.py]:
```suggestion
        class C:
            __attr = "_C__attr"
        class Outer:
            __attr = "_Outer__attr"
            def f(self, x):
                match x:
                    # looks up __attr, not _C__attr or _Outer__attr
                    case C(__attr=y):
                        return y
        c = C()
        c.__attr = "__attr"
        self.assertEqual(Outer().f(c), "__attr")
```

**Review by brandtbucher — APPROVED** [review]
Thanks for the test. I think the readability could be a bit improved a bit (it took me a second to figure out what this test was doing)... what do you think of this version?

If you don't like it, at the very least we should change all of the `setattr`s to normal attribute assignments.

File: Lib/test/test_patma.py [hunk 2]
```diff
@@ -2654,6 +2654,20 @@ def get(key, default=None):
...
+        class Outer:
+            __attr = "_Outer__attr"
+            def f(self, x):
+                match x:
+                    # looks up __attr, not _C__attr or _Outer__attr
+                    case C(__attr=y):
+                        return y
+        c = C()
+        setattr(c, "__attr", "__attr")  # setattr is needed because we're in a class scope
+        self.assertEqual(Outer().f(c), "__attr")
```
Comment (gvanrossum) [inline Lib/test/test_patma.py]: I'd use something other than `"__attr"` for the *value* here, to avoid being too cute.
```suggestion
        setattr(c, "__attr", "spam")  # setattr is needed because we're in a class scope
        self.assertEqual(Outer().f(c), "spam")
```

**Review by gvanrossum — COMMENTED** [review]
LG with one suggestion. I approve of locking in this behavior. @brandtbucher Do you think we should add a clarification to PEP 634?

**Comment by brandtbucher** [comment]
> Do you think we should add a clarification to PEP 634?

I think the current wording ("The keyword is looked up as an attribute on the subject.") is fine as-is, but I wouldn't reject a PR proposing to modify it.

Nobody really reads 634, though. 🙃

File: Lib/test/test_patma.py (hunk 2 above)
Comment (da-woods) [reply Lib/test/test_patma.py]: Done

File: Lib/test/_testcppext.cpp [hunk 3]
```diff
@@ -48,6 +48,31 @@ class StrongRef
...
+    "IsPyObject", // tp_name
+    sizeof(IsPyObject), // tp_basicsize
+    0, // tp_itemsize
+    IsPyObject::dealloc,  // tp_dealloc
+};
+
+IsPyObject::IsPyObject() {
+    ob_type = &ipotype;
+    Py_INCREF(this);
+}
```
Comment (encukou) [inline Lib/test/_testcppext.cpp]: You've included some unrelated code (the reproducer for #94731) in this PR.

Comment (da-woods) [reply Lib/test/_testcppext.cpp]: Thanks.... I obviously did `git commit -a` when I shouldn't have done. Sorry for the false alarm!

**Comment by da-woods** [comment]
Rebased to undo a git mishap! Sorry!

**Review by gvanrossum — APPROVED** [review]
Looks fine to me.

**Comment by gvanrossum** [comment]
Since it's just a test I don't feel that it needs to be backported, though we could probably backport to 3.11 and 3.10. @brandtbucher What do you think?

**Comment by brandtbucher** [comment]
Nah, I’m not worried about this behavior regressing on older branches.
